tools/lsp/common/document_cache.rs

```diff
@@ -0,0 +1,419 @@
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-Royalty-free-2.0 OR LicenseRef-Slint-Software-3.0
+
+//! Data structures common between LSP and previewer
+
+use i_slint_compiler::diagnostics::{BuildDiagnostics, SourceFile};
+use i_slint_compiler::object_tree::Document;
+use i_slint_compiler::typeloader::TypeLoader;
+use i_slint_compiler::typeregister::TypeRegister;
+use lsp_types::Url;
+
+use std::{
+    cell::RefCell,
+    collections::HashMap,
+    future::Future,
+    path::{Path, PathBuf},
+    pin::Pin,
+    rc::Rc,
+};
+
+use crate::common::{file_to_uri, uri_to_file, ElementRcNode, Result};
+
+pub type SourceFileVersion = Option<i32>;
+
+pub type SourceFileVersionMap = HashMap<PathBuf, SourceFileVersion>;
+
+fn default_cc() -> i_slint_compiler::CompilerConfiguration {
+    i_slint_compiler::CompilerConfiguration::new(
+        i_slint_compiler::generator::OutputFormat::Interpreter,
+    )
+}
+
+pub type OpenImportFallback = Option<
+    Rc<
+        dyn Fn(
+            String,
+        ) -> Pin<
+            Box<dyn Future<Output = Option<std::io::Result<(SourceFileVersion, String)>>>>,
+        >,
+    >,
+>;
+
+pub struct CompilerConfiguration {
+    pub include_paths: Vec<std::path::PathBuf>,
+    pub library_paths: HashMap<String, std::path::PathBuf>,
+    pub style: Option<String>,
+    pub open_import_fallback: OpenImportFallback,
+    pub resource_url_mapper:
+        Option<Rc<dyn Fn(&str) -> Pin<Box<dyn Future<Output = Option<String>>>>>>,
+}
+
+impl Default for CompilerConfiguration {
+    fn default() -> Self {
+        let mut cc = default_cc();
+
+        Self {
+            include_paths: std::mem::take(&mut cc.include_paths),
+            library_paths: std::mem::take(&mut cc.library_paths),
+            style: std::mem::take(&mut cc.style),
+            open_import_fallback: None,
+            resource_url_mapper: std::mem::take(&mut cc.resource_url_mapper),
+        }
+    }
+}
+
+impl CompilerConfiguration {
+    fn to_compiler_configuration(
+        mut self,
+    ) -> (i_slint_compiler::CompilerConfiguration, OpenImportFallback) {
+        let mut result = default_cc();
+        result.include_paths = std::mem::take(&mut self.include_paths);
+        result.library_paths = std::mem::take(&mut self.library_paths);
+        result.style = std::mem::take(&mut self.style);
+        result.resource_url_mapper = std::mem::take(&mut self.resource_url_mapper);
+
+        (result, self.open_import_fallback)
+    }
+}
+
+/// A cache of loaded documents
+pub struct DocumentCache {
+    type_loader: TypeLoader,
+    open_import_fallback: OpenImportFallback,
+    source_file_versions: Rc<RefCell<SourceFileVersionMap>>,
+}
+
+pub fn document_cache_parts_setup(
+    compiler_config: &mut i_slint_compiler::CompilerConfiguration,
+    open_import_fallback: OpenImportFallback,
+    initial_file_versions: SourceFileVersionMap,
+) -> (OpenImportFallback, Rc<RefCell<SourceFileVersionMap>>) {
```

Comment:
there is certainly a reason why you return the open_import_fallback that you just passed. but that is't really clear, what is the relation with the open_import_fallback in the self::CompilerConfiguration and in the DocumentCache? why do we need two?

Reply:
I need to be able to `snapshot` a `DocumentCache`: That snapshot needs to continue tracking version changes, but it must do so independent of the `DocumentCache` it was cloned from. So I need to keep the code that the `DocumentCache`s share to find the file contents and version separate from the tracking of which file was read in which version.

`OpenImportFallback` is the re-usable part that is used to find the contents of a file and its version information based on a file path. It will always return the latest contents and version info available.

The compiler's `i_slint_compiler::CompilerConfiguration` has the code wiring up the `OpenImportFallback` with the version information tracking and that is not reusable between the different `DocumentCache`s.